subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -0,0 +1,241 @@
+	res = res && zcbor_list_start_decode(states);
+	res = res && zcbor_multi_decode(1, CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT, &image_count,
+					(zcbor_decoder_t *)parse_image_info, states,
+					ctx.header.images, sizeof(struct image_info));
+	res = res && zcbor_list_end_decode(states);
+	res = res && zcbor_map_end_decode(states);
+
+	if (!res) {
+		return zcbor_pop_error(states);
+	}
+
+	ctx.header.image_count = image_count;
+	select_next_image();
+
+	return 0;
+}
+
+static const struct dfu_image_writer *current_image_writer(void)
+{
+	if (ctx.cur_image_no >= 0 && (size_t)ctx.cur_image_no < ctx.header.image_count) {
+		const int image_id = ctx.header.images[ctx.cur_image_no].id;
+
+		for (size_t i = 0; i < ctx.writer_count; i++) {
+			if (ctx.writers[i].image_id == image_id) {
+				return &ctx.writers[i];
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static int process_current_item(const uint8_t *chunk, size_t chunk_size)
+{
+	int err = 0;
+
+	/* Process only remaining bytes of the current item (header or image) */
+	chunk_size = MIN(chunk_size, ctx.cur_item_size - ctx.cur_item_offset);
+
+	if (ctx.cur_image_no < 0) {
+		/* Parsing header */
+		memcpy(ctx.buffer + ctx.cur_item_offset, chunk, chunk_size);
+		ctx.cur_item_offset += chunk_size;
+
+		if (ctx.cur_item_offset == ctx.cur_item_size) {
+			err = parse_header();
+		}
+	} else {
+		/* Image data */
+		const struct dfu_image_writer *writer = current_image_writer();
+
+		if (!writer) {
+			err = -ESPIPE;
+		}
+
+		if (!err && ctx.cur_item_offset == 0) {
+			err = writer->open(writer->image_id);
+		}
+
+		if (!err) {
+			err = writer->write(chunk, chunk_size);
+			ctx.cur_item_offset += chunk_size;
+		}
+
+		if (!err && ctx.cur_item_offset == ctx.cur_item_size) {
+			err = writer->close(true);
+			select_next_image();
+		}
+	}
+
+	if (err) {
+		return err;
+	}
+
+	return chunk_size;
+}
+
+int dfu_multi_image_init(uint8_t *buffer, size_t buffer_size, dfu_multi_image_seek_t seek)
+{
+	if (buffer == NULL || buffer_size < FIXED_HEADER_SIZE) {
+		return -EINVAL;
+	}
+
+	memset(&ctx, 0, sizeof(ctx));
+	ctx.buffer = buffer;
+	ctx.buffer_size = buffer_size;
+	ctx.seek = seek;
+	ctx.cur_image_no = -1;
+	ctx.cur_item_size = FIXED_HEADER_SIZE;
+
+	return 0;
+}
+
+int dfu_multi_image_register_writer(const struct dfu_image_writer *writer)
+{
+	if (ctx.writer_count == CONFIG_DFU_MULTI_IMAGE_MAX_IMAGE_COUNT) {
+		return -ENOMEM;
+	}
+
+	ctx.writers[ctx.writer_count++] = *writer;
+	return 0;
+}
+
+int dfu_multi_image_write(size_t offset, const uint8_t *chunk, size_t chunk_size)
+{
+	if (offset > ctx.cur_offset) {
+		/* Unexpected data gap */
+		return -ESPIPE;
+	}
+
+	/* Skip already processed bytes (if any) */
+	const size_t skip_size = MIN(ctx.cur_offset - offset, chunk_size);
+
+	chunk += skip_size;
+	chunk_size -= skip_size;
+
+	/* Mark the chunk as already processed */
+	ctx.cur_offset += chunk_size;
+
+	while (chunk_size > 0) {
+		int res = process_current_item(chunk, chunk_size);
```

Comment:
is there a case where this could return an error, and the solution would be to repeat the operation? 
if so, we should mark the offset (line 210) after this while loop, so that write could be repeated.

Reply:
The thing is if we implement the skipping logic `ctx.cur_offset` will need to be updated inside `process_current_item`, so I wanted to increase the variable before. I can probably move the code inside `process_current_item`, too. But is there any scenario in which you think a write could fail, but a subsequent write of the same image chunk would succeed? I thought it wasn't a realistic scnario.

Reply:
Ok, So in our case this will be correct.

About a case where another call after an error may fix the flow - I was thinking about the implications of the generic approach in this interface. By allowing injection of the writer function, we adopt the characteristics and limitations of that function. Current declaration is not very restrictive, so we can not expect how those custom writers will handle errors. 
So let's assume that we have a customer's writer. This could be a simple writer, that writes a chunk of data on flash. This function can return an error that would mean that the flash device is busy. In that case, calling the right again can result in correct operation. 

I can see two solutions to this dilemma, first one is to just add documentation to the function pointers, that will tell the user that this function may return error only in unrecoverable state, and returning without error means that the data has been correctly stored. 
Another approach would be to expect recoverable errors in the dfu_multi_image, but as You mentioned this increases the complexity. 

I also think that the restrictions of the function should be obvious just based on the function interface (name, parameters, return type, module conversion). Unfortunately, I do not have any clever idea how to write such interface in this case.

Reply:
I think the current documentation is clear that after you get an error from `dfu_multi_image_write` (which forwards any error code returned by an image writer's `write` function), you shall not write any more chunks. But OK, I can, at the cost of increasing some flash usage, reorganize the code, so that it can handle repeated writes.

Reply:
I changed the code, so that offsets are not updated on any error. Also extended the documentation significantly. Let me know if now it's clear.